### DISCUSSION TITLE
Moving point labels without recreating

### DISCUSSION THREAD

**Discussion by ljrobins** [pyvista/pyvista#4607] [Q&A] [answered]
I'm building a tool that visualizes orbits and I'd like the labels for celestial bodies and satellites to follow them as they move around. With a normal 3D actor, I can just set the `pl.actors[key].user_matrix` to the 4x4 of my choice, but there doesn't seem to be an equivalent method to reposition point labels.

I've dug into the `vtkActor2D` and `vtkProp` APIs and can't find anything for updating the anchor point. It feels like `pl.actors['my-labels'].GetMapper().GetAnchorTransform().SetValue(x, y, z)` should do something, but it does not.

This seems like something that pyvista should be able to support as most GUIs and scientific animations would benefit from being able to directly manipulate labels without deleting and recreating them each time.

**Comment by banesullivan** [comment, marked as answer]
Unfortunately, I think you'll need to change the actual anchor point location instead of applying a translation matrix (I could be wrong, though).

One approach (and there are a few) to do this would be:

```py
import numpy as np
import pyvista as pv

# Make points along a spline
theta = np.linspace(-1 * np.pi, 1 * np.pi, 100)
z = np.linspace(2, -2, 100)
r = z**2 + 1
x = r * np.sin(theta)
y = r * np.cos(theta)
points = np.column_stack((x, y, z))

# Create a spline. This automatically computes arc_length, which is the
# distance along the line.
spline = pv.Spline(points, 1000)

# Create a mesh that will hold the location which updates
label = pv.PolyData([spline.points[0]])
label.point_data['label'] = ['My Location']

pl = pv.Plotter()
pl.add_mesh(spline, color=True, render_lines_as_tubes=True, line_width=10)
pl.add_point_labels(
    label, 'label', point_color='red', always_visible=True, point_size=20, render_points_as_spheres=True
)
pl.show_bounds()
pl.show_axes()
pl.camera_position = 'xz'

# Animate
pl.open_movie('animation.mp4')
pl.write_frame()
for i in range(1, spline.n_points):
    label.points[0] = spline.points[i]
    pl.render()
    pl.write_frame()

pl.close()
```

https://github.com/pyvista/pyvista/assets/22067021/599dd069-a784-4b19-955e-74ca7c3ecd18

**Reply by ljrobins** [reply]
Awesome, I can live with that. I had tried something very simila (`label.translate(x, y, z)`), but since that's probably just updating the actor `user_matrix`, it didn't pull the label with. 

Ideally, `.translate` would be treated the same by the label actor, but moving the points directly is fine by me!

**Reply by banesullivan** [reply]
You could perhaps use `.translate(..., inplace=True)` on the `label` PolyData object above to get the same affect

**Reply by banesullivan** [reply]
Please note that there is a weird split in how input points are handled in `add_point_labels` such that this will only work if the label is a point data array on the source points:

https://github.com/pyvista/pyvista/blob/8adee29f0e7d08741e7a09190d502f9327419b63/pyvista/plotting/plotter.py#L5328-L5346

**Reply by ljrobins** [reply]
So it does! That's the cleanest solution. 

p.s., `pyvista` is my favorite thing since sliced bread it has singlehandedly allowed me to transition all my research code from MATLAB to Python. I'd love to give back over the next few years of my PhD, you might see me in the PRs soon enough.

**Reply by banesullivan** [reply]
Love to hear it!! We welcome any contributions you can give!
